Proposal

PolyProgrammist opened the pull request.

petersalomonsen commented:
Here's the issue to link to @PolyProgrammist  https://github.com/NEAR-DevHub/neardevhub-bos/issues/633

frol commented on src/lib.rs:
```diff
@@ -198,7 +222,46 @@ impl Contract {
+
+        let desc = get_proposal_description(proposal.clone());
+
+        proposal::repost::publish_to_socialdb_feed(proposal);
```
You need to chain a [callback](https://docs.near.org/sdk/rust/cross-contract/callbacks) here to capture the block height received from SocialDB.

frol reviewed (changes requested):
Please, add simulation tests. As example, see the community tests: https://github.com/NEAR-DevHub/neardevhub-contract/blob/main/tests/communities.rs

frol commented on src/lib.rs:
```diff
@@ -198,7 +223,75 @@ impl Contract {
+        assert_eq!(env::promise_results_count(), 1, "ERR_TOO_MANY_RESULTS");
+        match env::promise_result(0) {
+            PromiseResult::Successful(val) => {
+                if let Ok(set_result) = near_sdk::serde_json::from_slice::<Value>(&val) {
+                    let block_height: u64 = set_result["block_height"].as_str().unwrap().parse::<u64>().unwrap();
```
If you define the interface with the return value that is strictly typed and implement serde::Deserialize (see my other comment), you won't need to parse it manually here.

frol commented on src/lib.rs:
```diff
@@ -210,16 +303,35 @@ impl Contract {
+
     pub fn get_all_labels(&self) -> Vec<String> {
         near_sdk::log!("get_all_labels");
-        let mut res: Vec<_> = self.label_to_posts.keys().collect();
+        let mut post_labels: HashSet<_> = self.label_to_posts.keys().collect::<HashSet<_>>();
+        let proposals_labels: HashSet<_> = self.label_to_proposals.keys().collect::<HashSet<_>>();
+        post_labels.extend(proposals_labels);
+        let mut res: Vec<_> = post_labels.into_iter().collect();
         res.sort();
         res
     }
 
     pub fn get_all_authors(&self) -> Vec<AccountId> {
         near_sdk::log!("get_all_authors");
-        let mut res: Vec<_> = self.authors.keys().collect();
+        let mut author_posts: HashSet<_> = self.authors.keys().collect::<HashSet<_>>();
+        let author_proposals: HashSet<_> = self.author_proposals.keys().collect::<HashSet<_>>();
+        author_posts.extend(author_proposals);
+        let mut res: Vec<_> = author_posts.into_iter().collect();
         res.sort();
         res
     }
```
I suggest we don't touch these and the methods below, instead, create a separate set of methods for proposals.

frol commented on src/lib.rs:
```diff
@@ -644,6 +823,64 @@ mod tests {
+        }"#).unwrap();
+        contract.add_proposal(VersionedProposalBody::V0(body), HashSet::new());
+        let receipts = get_created_receipts();
+        println!("{:?}", receipts);
+        assert_eq!(3, receipts.len());
+        let receipt = receipts.get(2).unwrap();
+        let receipt_str = format!("{:?}", receipt);
+        let re = Regex::new(r#"method_name: (\[[^\]]*\]), args: (\[[^\]]*\])"#).unwrap();
+
+        // Extract the method_name and args values
+        for cap in re.captures_iter(&receipt_str) {
+            let method_name = &cap[1];
```
Hmm, what is going on here? It seems that you should be able to just:

```rust
if let near_sdk::mock::MockAction::FunctionCallWeight { method_name, args } = receipts[2].actions[0] {
    assert_eq!(method_name, b"set");
    assert_eq!(args, b"...");
} else {
    assert!(false, "Expected a function call ...")
}
```

frol commented on src/lib.rs:
```diff
@@ -198,7 +223,75 @@ impl Contract {
+            .with_static_gas(env::prepaid_gas().saturating_div(3))
+            .set_block_height_callback(proposal.clone()), proposal.clone());
+
+        let desc = get_proposal_description(body);
+
+        notify::notify_mentions(desc.as_str(), id);
```
This looks like we need to refactor to be less hacky. We need to notify the sponsor, supervisor (optionally), and anyone mentioned in the proposal summary and description except the author themselves.

frol commented on src/lib.rs:
```diff
@@ -198,7 +223,75 @@ impl Contract {
+
+
+    #[private]
+    pub fn set_block_height_callback(&mut self, old_proposal: Proposal) -> u64 {
```
I think we should return `{"proposal_id": "u64-stringified value - near_sdk::json_type::U64"}`

frol commented on src/lib.rs:
```diff
@@ -198,7 +223,75 @@ impl Contract {
+            PromiseResult::Successful(val) => {
+                if let Ok(set_result) = near_sdk::serde_json::from_slice::<Value>(&val) {
+                    let block_height: u64 = set_result["block_height"].as_str().unwrap().parse::<u64>().unwrap();
+                    let proposal = Proposal {
+                        id: old_proposal.id,
+                        author_id: old_proposal.author_id.clone(),
+                        block_height: block_height,
+                        snapshot: old_proposal.snapshot,
+                        snapshot_history: old_proposal.snapshot_history,
+                    };
```
You can define `old_proposal` as `mut` right in the function signature (since you expect the parameter to be passed "by value", it means that you will own it, and so it is totally fine to also mark it as internally mutable)

```rust
pub fn set_block_height_callback(&mut self, mut old_proposal: Proposal)
...

old_proposal.block_height = block_height;
self.proposals....
```

frol reviewed (changes requested):
Please, think about naming of fields from the user perspective. E.g. `block_height` inside Proposal struct does not make a ton of sense, we should name it something like `social_db_post_block_height`

PolyProgrammist commented:
TODO:
 - Add the ability to change categories for the admin
 - Check that when adding a proposal it is in draft status
 - Check that when adding there cannot be payouts
 - Check that regular user can only edit draft
 - Check whether timeline_status was done correctly
 - Send notifications with correct type & value
 - Add notifications to edit_post
 - Create test for payouts and maybe change type

PolyProgrammist commented:
TODO:
Change payouts type
Change timeline status json
Check notifications don't break anything
Rebase
Check everything once again and line up with frontend

ailisp reviewed (commented):
I think this looks great! It should not affect or break existing functionality of production contract.
Please deploy a copy to testnet so frontend development can happen at same time.

PolyProgrammist commented:
> I think this looks great! It should not affect or break existing functionality of production contract. Please deploy a copy to testnet so frontend development can happen at same time.

Hi, I have posted information about the address of the deployed contract as well as info about how to call it's functions here https://github.com/NEAR-DevHub/neardevhub-contract/issues/89#issuecomment-1924182487

PolyProgrammist commented:
I have created a pull request in original repository, let's move discussions there https://github.com/NEAR-DevHub/neardevhub-contract/pull/90

PolyProgrammist committed:
New Proposal workflow

PolyProgrammist committed:
Review:
Return Promise for add_proposal
Methods for timeline
ProposalId to u32
Replaced Balance in proposal for u32, moved Balance from lib.rs to post/mod.rs

PolyProgrammist committed:
Review:
Avoid try_into where possible
Use ProposalId in BlockHeightCallbackRetValue
Change link for socialdb
Fixed StorageKey enum

PolyProgrammist committed:
Fix link for post

PolyProgrammist committed:
Fix tests for post link

PolyProgrammist committed:
Review: inline get_proposal_summary

PolyProgrammist committed:
Remove get_proposal_summary
